source/TimePicker.js

```diff
@@ -585,14 +608,27 @@
 		meridiemPickerChanged: function (sender, event) {
 			if(this.syncingPickers) return true;
 
-			var hour = this.$.hour.get('value'),
-				value = event.value;
+			var meridiems = this.meridiems,
+				hour = this.$.hour.getValue(),
+				oldIndex = event.old,
+				newIndex = this.$.meridiem.getValue(),
+				offset, start, end, newHour;
 
-			if (this.value) {
-				// value is 0 for am, 1 for pm
-				// reset the hour to < 12 and then add 12 if it's pm
-				hour = hour%12 + value*12;
-				this.updateHours(hour);
+			offset = hour - parseInt(meridiems[oldIndex]['start'], 10);
+			start = parseInt(meridiems[newIndex]['start'], 10);
+			end = parseInt(meridiems[newIndex]['end'], 10);
+			newHour = offset + start;
+
+			if (newHour != hour) {
+				if (newHour > end) {
+					newHour = end;
+				}
+				this.updateValue('hour', this.value.getHours() + (newHour - hour));
```

Comment:
It seems like this triggers a value update for `TimePicker`, which eventually calls `setChildPickers` and sets the meridiem value back to the previous value.